fqdn: Update high-level package docs (#11034)
cilium-team/cilium: Documentation/contributing/development/debugging.rst

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    http://docs.cilium.io

###############
Troubleshooting
###############

toFQDNs and DNS Debugging
-------------------------

The interactions of L3 toFQDNs and L7 DNS rules can be difficult to debug
around. Unlike many other policy rules, these are resolved at runtime with
unknown data. Pods may create large numbers of IPs in the cache or the IPs
returned are not compatible with our datapath implementation. Sometimes we also
just have bugs.

Isolating the source of toFQDNs issues
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

While there is no common culprit when debugging, the DNS Proxy is the most
one-off component in this chain. The cascading caching scheme is also complex

> in its behaviour. Isolating an issue to these components, or to the policy

Reviewer: > Isolating an issue to these components

to

> Isolating an issue **in** these components

layer or datapath is often the first step when debugging toFQDNs related
issues. Generally, working top-down is easiest as the information needed to
verify low-level correctness can be collected in the initial debug invocations.

*REFUSED vs  NXDOMAIN responses*
The proxy uses REFUSED DNS responses to indicate a denied request. Some libc
implementations, notably musl which is common in alpine images, terminate the
whole DNS search in these cases. To work around it, denied responses can be

> configured to be NXDOMAIN via ``--tofqdns-dns-reject-response-code``.

Reviewer: change

> via

to

> by setting the command line argument

*Monitor Events*

> The DNS Proxy emits L7 monitor events. One for the request and one for the

Reviewer: Assuming I understand, I would reword to this:

The DNS Proxy emits multiple L7 monitor events. One for the request and one for the response (if allowed). Often the L7 DNS rules are paired with L3 toFQDNs rules and, thus, emits extra events.

response, if allowed. Often the L7 DNS rules are paired with L3 toFQDNs rules
and events are emitted. 

.. Note::

>     Bu sure to run cilium monitor on the same node as the pod being debugged!

Reviewer: Be

  request was allowed (based on Endpoint ID, destination IP:Port and the DNS
  query) an error occurred. These errors would come from the internal rule
  lookup in the proxy, the ``allowed`` field.
- ``Timeout waiting for response to forwarded proxied DNS lookup``: The proxy
  forwards requests 1:1 and does not cache. It applies a 5s timeout on
  responses to those requests, as the client will retry within this period
  (usually). Bursts of these errors can happen if the DNS target server
  misbehaves and many pods see DNS timeouts. This isn't an actual problem with
  cilium or the proxy although it can be caused by policy blocking the DNS
  target server if it is in-cluster.
- ``Timed out waiting for datapath updates of FQDN IP information; returning
  response``: When the proxy updates the DNS caches with response data, it
  needs to allow some time for that information to get into the datapath.
  Otherwise, pods would attempt to make the outbound connection (the thing that
  caused the DNS lookup) before the datapath is ready. Many stacks retry the
  SYN in such cases but some return an error and some apps further crash as a

>   response. This delay is configurable via

Reviewer: by setting the command argument

Note that the labels for identities are resolved here. This can be skipped, or there may be cases where this doesn't occur:

.. code:: bash

    $ kubectl exec pod/cilium-sbp8v -n cilium -- cilium bpf policy get -n 3459
    DIRECTION   IDENTITY   PORT/PROTO   PROXY PORT   BYTES   PACKETS
    Ingress     0          ANY          NONE         1367    7
    Ingress     1          ANY          NONE         0       0
    Egress      0          53/TCP       36447        0       0
    Egress      0          53/UDP       36447        138     2
    Egress      16777217   ANY          NONE         477     6

L3 ``toFQDNs`` rules are egress only, so we would expect to see an ``Egress``
entry with Security Identity ``16777217``. The L7 rule, used to redirect to the
DNS Proxy is also present with a populated ``PROXY PORT``. It has a 0
``IDENTITY`` as it is an L3 wildcard.

An identity missing here can be an error in various places:
- Policy doesn't actually allow this Endpoint to connect. A sanity check is to
use ``cilium endpoint list`` to see if cilium thinks it should have policy
enforcement.
- Endpoint regeneration is slow and the Policy Map has not been updated yet.

> This can occur in cases where we have leaked IPs from the DNS cache (i.e. the

Reviewer: they were never

were never deleted correctly) or when there are legitimately many IPs. It can

> also simply mean an overloaded node or even a deadlock within cilium.

Reviewer: In this case, is the deadlock in cilium you are suggesting the result of resource constraints?

Author: no! an actual lock vs lock deadlock. We have had those because we have a bunch of concurrent stuff that synchronises with a combination of locks, channels and wait groups so it can get messy.

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:

>     http://docs.cilium.io

Reviewer 2: ```suggestion
    https://docs.cilium.io
```
We switched to the `https` version for this URL in all the documentation a few days ago

#########
Debugging
#########

toFQDNs and DNS Debugging
-------------------------

The interactions of L3 toFQDNs and L7 DNS rules can be difficult to debug
around. Unlike many other policy rules, these are resolved at runtime with
unknown data. Pods may create large numbers of IPs in the cache or the IPs

> returned are not compatible with our datapath implementation. Sometimes we also

Reviewer 2: ```suggestion
returned may not be compatible with our datapath implementation. Sometimes we also
```

just have bugs.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Isolating the source of toFQDNs issues
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

While there is no common culprit when debugging, the DNS Proxy is the most

> one-off component in this chain. The cascading caching scheme is also complex

Reviewer 3: I'm not quite sure what `the most one-off component` means.

Author: I had "most likely to have bugs" in mind but I changed the wording to say that.

in its behaviour. Determining whether an issue is caused by the DNS components,
in the policy layer or in the datapath is often the first step when debugging
toFQDNs related issues. Generally, working top-down is easiest as the
information needed to verify low-level correctness can be collected in the
initial debug invocations.

REFUSED vs NXDOMAIN responses
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The proxy uses REFUSED DNS responses to indicate a denied request. Some libc

> implementations, notably musl which is common in alpine images, terminate the

Reviewer 3: ```suggestion
implementations, notably musl which is common in Alpine Linux images, terminate the
```

> whole DNS search in these cases. To work around it, denied responses can be

Reviewer 3: What is the user-facing error in these cases? Could we provide an example of what this looks like?

Author: It's app dependent, I'm afraid. The return from the lookup function is empty and the app would then error. I added a bit more explanation of what to look for.

    -> 0: 10.60.0.182:41510 -> 104.198.14.52:80 tcp ACK
    -> stack flow 0x614e9723 identity 323->16777217 state established ifindex 0 orig-ip 0.0.0.0: 10.60.0.182:41510 -> 104.198.14.52:80 tcp ACK, FIN
    -> 0: 10.60.0.182:41510 -> 104.198.14.52:80 tcp ACK, FIN
    -> endpoint 3459 flow 0x7388921 identity 16777217->323 state reply ifindex lxc84b58cbdabfe orig-ip 104.198.14.52: 104.198.14.52:80 -> 10.60.0.182:41510 tcp ACK, FIN
    -> stack flow 0x614e9723 identity 323->16777217 state established ifindex 0 orig-ip 0.0.0.0: 10.60.0.182:41510 -> 104.198.14.52:80 tcp ACK

The above is for a simple ``curl cilium.io`` in a pod. The L7 DNS request is
the first set of message and the subsequent L3 connection is the HTTP
component. AAAA DNS lookups commonly happen but were removed to simplify the
example.

- If no L7 DNS requests appear, the proxy redirect is not in place. This may
  mean that the policy does not select this endpoint or there is an issue with
  the proxy redirection.
  In the past, a bug occurred with more permissive L3 rules overriding the

>   proxy redirect, causing the proxy to never see the requests.

Reviewer 3: Note, users can check whether there are redirects using `cilium status --all-redirects`.

- If the L7 DNS request is blocked, with an explicit denied message, then the

>   requests are not allowed by the proxy. This may be due to a typo or the

Reviewer 3: ```suggestion
  requests are not allowed by the proxy. This may be due to a typo in the network policy, or the
```

  matchPattern rule not being correct. It can also be due to a bug in policy

>   propagation to the DNS Proxy.

Reviewer 3: Perhaps, "If you believe it is a bug, please <file an issue>"?

Author: This was more to allude to the section related to this. There's some basic problem isolation around the policy part later.

- If the DNS request is allowed, with an explicit message, and it should not
  be, this may be because a more general policy is in place that allows the
  request. ``matchPattern: "*"`` visibility policies are commonly in place and
  would supersede all other, more restrictive, policies.
  If no other policies are in place, incorrect allows may indicate a bug when
  passing policy information to the proxy. There is no way to dump the rules in
  the proxy, but a debug log is printed when a rule is added. Look for 
  ``DNS Proxy updating matchNames in allowed list during UpdateRules``.

>   See pkg/proxy/dns.go

Reviewer 3: This last fragment doesn't seem fully formed. Maybe..

```suggestion
  The pkg/proxy/dns.go file contains the DNS proxy implementation.
```

- ``Timeout waiting for response to forwarded proxied DNS lookup``: The proxy
  forwards requests 1:1 and does not cache. It applies a 5s timeout on
  responses to those requests, as the client will retry within this period
  (usually). Bursts of these errors can happen if the DNS target server
  misbehaves and many pods see DNS timeouts. This isn't an actual problem with
  cilium or the proxy although it can be caused by policy blocking the DNS
  target server if it is in-cluster.
- ``Timed out waiting for datapath updates of FQDN IP information; returning
  response``: When the proxy updates the DNS caches with response data, it
  needs to allow some time for that information to get into the datapath.
  Otherwise, pods would attempt to make the outbound connection (the thing that
  caused the DNS lookup) before the datapath is ready. Many stacks retry the
  SYN in such cases but some return an error and some apps further crash as a
  response. This delay is configurable by setting the
  ``--tofqdns-proxy-response-max-delay`` command line argument but defaults to
  100ms. It can be exceeded if the system is under load.

Identities and Policy
~~~~~~~~~~~~~~~~~~~~~

> Once a DNS response has been passed back through the proxy and is places in the

Reviewer 4: "placed"

DNS cache ``toFQDNs`` rules can begin using the IPs in the cache. There are
multiple layers of cache:

- A per-Endpoint ``DNSCache`` stores the lookups for this endpoint. It is
  restored on cilium startup with the endpoint. Limits are applied here for
  ``--tofqdns-endpoint-max-ip-per-hostname`` and TTLs are tracked. The

>   ``--tofqdns-min-ttl`` is not used here

Reviewer 3: ```suggestion
  ``--tofqdns-min-ttl`` is not used here.
```

- A per-Endpoint ``DNSZombieMapping`` list of IPs that have expired from the
  per-Endpoint cache but are waiting for the Connection Tracking GC to mark
  them in-use or not. This can take up to 12 hours to occur. This list is
  size-limited by ``--tofqdns-max-deferred-connection-deletes``. 
- A global ``DNSCache`` where all endpoint and poller DNS data is collected. It
  does apply the ``--tofqdns-min-ttl`` value but not the
  ``--tofqdns-endpoint-max-ip-per-hostname`` value.

If an IP exists in the FQDN cache (check with ``cilium fqdn cache list``) then
``toFQDNs`` rules that select a domain name, either explicitly via

> ``matchName`` or via ``matchPattern`` should cause IPs for that domain to have

Reviewer 2: ```suggestion
``matchName`` or via ``matchPattern``, should cause IPs for that domain to have
```

allocated Security Identities. These can be listed with:

.. code:: bash

    $ kubectl exec pod/cilium-sbp8v -n cilium -- cilium identity list
    ID         LABELS
    1          reserved:host
    2          reserved:world
    3          reserved:unmanaged
    4          reserved:health
    5          reserved:init
    6          reserved:remote-node
    323        k8s:class=xwing
               k8s:io.cilium.k8s.policy.cluster=default
               k8s:io.cilium.k8s.policy.serviceaccount=default
               k8s:io.kubernetes.pod.namespace=default
               k8s:org=alliance
    ...
    16777217   cidr:104.198.14.52/32
               reserved:world

> Note that CIDR identities are allocated locally on the node and have a high-bit set so they are often in the 16 Million range.

Reviewer 2: ```suggestion
Note that CIDR identities are allocated locally on the node and have a high-bit set so they are often in the 16-million range.
```

Also, break the long line for consistency?

                                                                                                                             6188
                                                                                                                             15194
                                                                                                                             18892
                                                                                                                             25379
                                                                                                                             29200
                                                                                                                             32255
                                                                                                                             33831
                                                                                                                             16777217
    &LabelSelector{MatchLabels:map[string]string{reserved.none: ,},MatchExpressions:[]LabelSelectorRequirement{},}   1

> In this example 16777217 is used by two selectors, one with ``matchPattern: "*"`` and another empty one. This is because of the policy in use:

Reviewer 2: Break the line?

.. code:: yaml

    apiVersion: cilium.io/v2
    kind: CiliumNetworkPolicy
    metadata:
      name: "tofqdn-dns-visibility"
    spec:
      endpointSelector:
        matchLabels:
          any:org: alliance
      egress:
      - toPorts:
          - ports:
             - port: "53"
               protocol: ANY
            rules:
              dns:
                - matchPattern: "*"
      - toFQDNs:
          - matchPattern: "*"

The L7 DNS rule has an implicit L3 allow-all because it defines only L4 and L7
sections. This is the second selector in the list, and includes all possible L3
identities known in the system. In contrast, the first selector, which
corresponds to the ``toFQDNS: matchName: "*"`` rule would list all identities
for IPs that came from the DNS Proxy. Other CIDR identities would not be
included.

> *Datapath Plumbing*

Reviewer 2: Does Sphinx render this as a paragraph title? I'd expect it to joint it with the next line.

Author: Nope! I had placeholders for the sections and I forgot to switch this one. Good catch!

For a policy to be fully realized the datapath for an Endpoint must be updated.
In the case of a new DNS-source IP, the CIDR identity associated with it must
propagate from the selectors to the Endpoint specific policy. Unless a new
policy is being added, this often only involves updating the Policy Map of the
Endpoint with the new CIDR Identity of the IP. This can be verified:

.. code:: bash

    $ kubectl exec pod/cilium-sbp8v -n cilium -- cilium bpf policy get 3459
    DIRECTION   LABELS (source:key[=value])   PORT/PROTO   PROXY PORT   BYTES   PACKETS
    Ingress     reserved:unknown              ANY          NONE         1367    7
    Ingress     reserved:host                 ANY          NONE         0       0
    Egress      reserved:unknown              53/TCP       36447        0       0
    Egress      reserved:unknown              53/UDP       36447        138     2
    Egress      cidr:104.198.14.52/32         ANY          NONE         477     6
                reserved:world 

> Note that the labels for identities are resolved here. This can be skipped, or there may be cases where this doesn't occur:

Reviewer 2: Break the line?

.. code:: bash

    $ kubectl exec pod/cilium-sbp8v -n cilium -- cilium bpf policy get -n 3459
    DIRECTION   IDENTITY   PORT/PROTO   PROXY PORT   BYTES   PACKETS
    Ingress     0          ANY          NONE         1367    7
    Ingress     1          ANY          NONE         0       0
    Egress      0          53/TCP       36447        0       0
    Egress      0          53/UDP       36447        138     2
    Egress      16777217   ANY          NONE         477     6

L3 ``toFQDNs`` rules are egress only, so we would expect to see an ``Egress``
entry with Security Identity ``16777217``. The L7 rule, used to redirect to the
DNS Proxy is also present with a populated ``PROXY PORT``. It has a 0

> ``IDENTITY`` as it is an L3 wildcard.

Reviewer 3: ```suggestion
``IDENTITY`` as it is an L3 wildcard, ie the policy allows any peer on the specified port.
```
